### PR TITLE
feat(search-option): add ticket age search option

### PR DESCRIPTION
**Pull request by ccailly**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 

## Description

This PR introduces a new 'Age' search option for tickets. This option is not usable for searching but only displayable in the results. The 'Age' is calculated as the time difference between the ticket's date and the current time. If a calendar is associated with the ticket, it calculates the active time between the ticket's date and the current time.

## Changes

- Added 'Age' search option to tickets.
- Updated `getSpecificValueToDisplay` method to handle the new 'Age' field.
- Added `ageSearchOptionDataProvider` method to provide data for testing the 'Age' option.
- Added `testAgeSearchOption` method to test the 'Age' option.

### CODE REVIEW

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Looks good, just two remarks that are up for discussion :

1\) "Age" was the name of the label used in the original specific patch of the client but I don't really like it. Maybe something like "Processing time" instead ? 

2\) I am not sure we should display the amount in days:

image

Since the time are computed base on a calendar where a day might only be 8 hours, it feels weird to show the total divided by 24 hours days.
I would suggest keeping only hours and minutes.

What are your thoughts on these two points @orthagh ?

**Comment by orthagh** [comment]
> "Processing time"

"Time since opening"

> amount in days

Since we present the delay with the calendar involved, I guess it's ok like done currently

**Review by orthagh — APPROVED** [review]

File: src/Ticket.php [hunk 1]
```diff
@@ -3277,6 +3289,22 @@ public static function getSpecificValueToDisplay($field, $values, array $options
...
+                    $calendar = new Calendar();
+                    $calendar->getFromDB($calendars_id);
+                    $time = $calendar->getActiveTimeBetween($ticket->fields['date'], $_SESSION["glpi_currenttime"]);
+                } else {
+                    $ticket_date = new DateTime($ticket->fields['date']);
+                    $now = new DateTime($_SESSION["glpi_currenttime"]);
+                    $time = $now->getTimestamp() - $ticket_date->getTimestamp();
+                }
+
+                return sprintf(__('%s days %s hours %s minutes'), floor($time / 86400), floor(($time % 86400) / 3600), floor(($time % 3600) / 60));
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply src/Ticket.php:3307]: I've made the same remark but @orthagh said it was fine with this format.
I agree that I would prefer without the days.

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Can you not display the days from the results (if OK for @orthagh) ?

File: src/Ticket.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (orthagh) [reply src/Ticket.php:3307]: I didn't saw the problem initialy, sorry.
If we want an age for a ticket, computed from the start date, we don't take into account calendars at all, no ?
Or the column title should mention it displays in working days but I guess it's still strange.

File: src/Ticket.php [hunk 2]
```diff
@@ -2701,6 +2701,17 @@ public function rawSearchOptions()
...
 
+        $tab[] = [
+            'id'                 => '70',
+            'table'              => $this->getTable(),
+            'field'              => '_virtual_age',
+            'datatype'           => 'specific',
+            'name'               => __('Time since opening'),
+            'massiveaction'      => false,
+            'nosearch'           => true,
+            'additionalfields'   => ['id']
```
Comment (AdrienClairembault) [inline src/Ticket.php:2712]: You can add `entities_id` and `date` into this array to fetch all the data you need at once.
You will then no longer need to reload the ticket from the database in the `getSpecificValueToDisplay` function.

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Meant to request change, not approve.

File: src/Ticket.php [hunk 3]
```diff
@@ -3277,6 +3290,20 @@ public static function getSpecificValueToDisplay($field, $values, array $options
...
+                    $calendar = new Calendar();
+                    $calendar->getFromDB($calendars_id);
+                    $time = $calendar->getActiveTimeBetween($values['date'], $_SESSION["glpi_currenttime"]);
+                } else {
+                    $ticket_date = new DateTime($values['date']);
+                    $now = new DateTime($_SESSION["glpi_currenttime"]);
+                    $time = $now->getTimestamp() - $ticket_date->getTimestamp();
+                }
+
+                return sprintf(__('%s hours %s minutes ago'), floor($time / 3600), floor(($time % 3600) / 60));
```
Comment (cedric-anne) [inline src/Ticket.php:3306]: The `ago` does not seems relevant here. Indeed, we are rendering a duration that does not reflect the real time, so we cannot say "16 hours ago" for a ticket created 2 days ago on a calendar that has 8 working hours per day.

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**